wolfcrypt/src/asn.c

```diff
@@ -4547,18 +4547,20 @@ static int GetCertHeader(DecodedCert* cert)
     if (GetSequence(cert->source, &cert->srcIdx, &len, cert->maxIdx) < 0)
         return ASN_PARSE_E;
 
+    /* Reset the max index for the size indicated in the outer wrapper. */
+    cert->maxIdx = len + cert->srcIdx;
     cert->certBegin = cert->srcIdx;
 
     if (GetSequence(cert->source, &cert->srcIdx, &len, cert->maxIdx) < 0)
         return ASN_PARSE_E;
     cert->sigIndex = len + cert->srcIdx;
 
     if (GetExplicitVersion(cert->source, &cert->srcIdx, &cert->version,
-                                                              cert->maxIdx) < 0)
+                                                            cert->sigIndex) < 0)
```

Comment:
Why sigIndex?  What if it's past maxIndex?

Reply:
The caller of GetCertHeader() is wc_GetPubX509(), which is an exported API. I was assuming that sigIndex would be set and already checked for being before maxIndex. I see that the call to wc_GetPubX509() from ssl.c will not have sigIndex set. Revising.

Reply:
I reviewed my changes and I believe they are correct as is. sigIndex is set to the end of the TBSCertificate sequence, which will be before maxIdx. (Otherwise, the GetSequence() on line 4554 would have returned a parse error.) My goal was to limit the scope of the data copies to within the enclosing sequence of a particular item, not to the end of the whole certificate, maxIdx.

Reply:
I believe the change is correct. sigIndex is set right after the call to GetSequence() on line 4554 where we are getting the bound on the sequence enclosing the TBSCertificate; the signature being the next item after. If sigIndex would have been beyond maxIdx, we should have returned with a parse error. My goal is to check that a parsed item exists within its immediate container sequence rather than the whole certificate buffer.

Reply:
What if len (as an int) ends up a negative number through a problem and the conversion to unsigned creates a new very big max end in sigIndex.  How about adding a check that we're still <= to maxIdx?

Reply:
Fair enough.